### ISSUE TITLE
incompatibility with async plugins

### ISSUE THREAD

**Issue by anru** [css-modules/css-modules-require-hook#58]
If you try to use require hook with any async plugin (for example `postcss-import`) you get error `throw new Error('Use process(css).then(cb) to work with async plugins');`

**Comment by sullenor** [comment]
@anru hi, I'm not sure that it's currently possible to support async plugins. The `require` function is synchronous, so I have to make an synchronous implementation as well.

**Comment by anru** [comment]
Well, there is not much problem, actually. In my case I just generate css modules map file in build time and use own hook that uses that map, so, may be a couple of words about it in Readme would be helpfull for someone else :)

**Comment by sullenor** [comment]
@anru yeah, looks like that question becomes more and more popular, so I'll add it to the readme.

Can you tell me, what plugins you use with require hook and why? Usually you need to use require hook to generate tokens and exception may occur if you use non-css files.

**Comment by jedrichards** [comment]
I'm using `postcss-import` to do something like,

```
@import "../colors.css";

.foo {
  color: var(--my-col);
}
```

And I get the async plugins error ...

**Comment by jedrichards** [comment]
@anru

> In my case I just generate css modules map file at build time and use own hook that uses that map

Can you expand on this a bit? Thanks :)

**Comment by jedrichards** [comment]
Btw, I'm making some progress with https://github.com/istarkov/babel-plugin-webpack-loaders

I think it removes the need for this require-hook by preprocessing your required modules synchronously with your Webpack loaders in a Node env ...

**Comment by sullenor** [comment]
@jedrichards Hi, thank you for your interest.
Actually I'm not sure, that there is an easy and performant solution to make async code work synchronously without monkey patching or whole js module compilation, but I'll look for the possibilities.

**Comment by anru** [comment]
@jedrichards 

1. First, you generate map file at build time, and save it to `css-modules.json` for example
2. Second, run small hook which uses that map for returning tokens for css files:

```js
  require.extensions['.css'] = function(m, filename) {
    const map = require('css-modules.json');
    const relativePath = path.relative(process.cwd(), filename);

    const tokens = map[relativePath];
    return m._compile(`module.exports = ${JSON.stringify(tokens)}`, filename);
  };
```
Thats all

**Comment by sullenor** [comment]
@anru nice solution :)
